webserver/webserver/src/main/java/io/helidon/webserver/HttpInitializer.java

```diff
@@ -60,7 +63,21 @@
     private final NettyWebServer webServer;
     private final SocketConfiguration soConfig;
     private final Routing routing;
-    private final Queue<ReferenceHoldingQueue<DataChunk>> queues = new ConcurrentLinkedQueue<>();
+    private final AtomicBoolean clearLock = new AtomicBoolean();
+
+    /**
+     * Reference queue that collects ReferenceHoldingQueue's when they become
+     * ready for garbage collection. ReferenceHoldingQueue's extracted from
+     * this collection that cannot be fully released (some buffers still in
+     * use) will be added to {@code unreleasedQueues} for later retries.
+     */
+    private final ReferenceQueue<Object> queues = new ReferenceQueue<>();
```

Comment:
Just being careful: is `Object` the type argument you want to pass here?

Reply:
Correct. That type-parameter is silly. (Reads: I don't get it) It is meant to represent the type of element returned by `References` enqueued in `ReferenceQueue`. But by the time a reference is enqueued the element is no longer reachable, as determined by GC. So...

What would make more sense as a type parameter for the `ReferenceQueue`, is the class of Reference that is going to be returned by `poll`.

Reply:
Well, I don't know about silly.  Would `? super Reference<?>` or something like that work better here?

Reply:
No.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/ReferenceQueue.html#poll() - `ReferenceQueue<T>.poll()` returns `Reference<? extends T>`. I am not interested in what `T` is, because that value is no longer accessible. I am only interested in the type of `Reference`. If it is something that has a method to release whatever resource it is associated with (ie `instanceof IndirectReference` in this case), I am going to call it - end of story.

I'd rather have a guarantee that I will not encounter other type of `Reference` here, so that I didn't have to type-cast. :)

Reply:
Think of it this way: `IndirectReference` is about two things: some owner of type `T` that actually does not matter to a resource `R` that is temporarily handed off to the owner of type `T`. The owner has the obligation to always terminate in a predictable path where it hands off the resource `R` back to some resource pool.

`IndirectReference` then is a safety net catching the cases when the owner `T` fails to fulfil the obligation, and dies before returning resource `R`. In this case `IndirectReference` ends up in the `ReferenceQueue`, and the queue processing routine is able to return resource `R` back to the pool. Note only the owner of type `T` is referenced phantomly. The resource `R` remains strongly reachable, at least through `IndirectReference` - unless, of course, the owner does fulfil the obligation, and returns the resource, first reclaiming it through `IndirectReference.acquire()`.

Reply:
Having explained all this, I think more work is needed.

The focus has been to reduce the cost of maintaining resources for well-terminating responses. But need to consider what happens to strong references when the channel is closed (`channelInactive` fired) and there are unfinished responses.